ml-agents/mlagents/trainers/trainer_controller.py

```diff
@@ -233,40 +231,55 @@ def start_learning(
                 trainer.write_tensorboard_text("Hyperparameters", trainer.parameters)
         try:
-            curr_info = self._reset_env(env)
+            for brain_name, trainer in self.trainers.items():
+                env_manager.set_policy(brain_name, trainer.policy)
+            self._reset_env(env_manager)
             while (
                 any([t.get_step <= t.get_max_steps for k, t in self.trainers.items()])
                 or not self.train_model
             ):
-                new_info = self.take_step(env, curr_info)
-                self.global_step += 1
-                if (
-                    self.global_step % self.save_freq == 0
-                    and self.global_step != 0
-                    and self.train_model
-                ):
-                    # Save Tensorflow model
-                    self._save_model(steps=self.global_step)
-                curr_info = new_info
+                n_steps = self.take_step(env_manager)
+                for i in range(n_steps):
+                    global_step += 1
+                    if (
+                        global_step % self.save_freq == 0
+                        and global_step != 0
+                        and self.train_model
+                    ):
+                        # Save Tensorflow model
+                        self._save_model(steps=global_step)
+                    for brain_name, trainer in self.trainers.items():
+                        # Write training statistics to Tensorboard.
+                        delta_train_start = time() - self.training_start_time
+                        if self.meta_curriculum is not None:
+                            trainer.write_summary(
+                                global_step,
+                                delta_train_start,
+                                lesson_num=self.meta_curriculum.brains_to_curriculums[
+                                    brain_name
+                                ].lesson_num,
+                            )
+                        else:
+                            trainer.write_summary(global_step, delta_train_start)
             # Final save Tensorflow model
-            if self.global_step != 0 and self.train_model:
-                self._save_model(steps=self.global_step)
+            if global_step != 0 and self.train_model:
+                self._save_model(steps=global_step)
         except KeyboardInterrupt:
             if self.train_model:
-                self._save_model_when_interrupted(steps=self.global_step)
+                self._save_model_when_interrupted(steps=global_step)
             pass
-        env.close()
+        env_manager.close()
         if self.train_model:
             self._write_training_metrics()
             self._export_graph()
 
-    def take_step(
-        self, env: BaseUnityEnvironment, curr_info: AllBrainInfo
-    ) -> AllBrainInfo:
+    def take_step(self, env: SubprocessEnvManager) -> int:
```

Comment:
Maybe rename to "advance" since it might take multiple steps?

Reply:
👍